File: python3/FindandReplacePattern/890.py
class Solution:
    def findAndReplacePattern(self, words: List[str], pattern: str) -> List[str]:
        def match(word):
            m1, m2 = {}, {}
            for w, p in zip(word, pattern):
                print(w, p)
                if w not in m1:
                    m1[w] = p
                if p not in m2:
                    m2[p] = w
                    
                #print(m1, m2)
                if (m1[w], m2[p]) != (p, w):
                    return False
            
            return True

        return filter(match, words)

File: go/FindandReplacePattern/890.go
package FindandReplacePattern

func findAndReplacePattern(words []string, pattern string) []string {
	var result []string
	var check bool = true

	for _, value := range words {
		m1 := make(map[string]string)
		m2 := make(map[string]string)
		for j, pat := range pattern {

			if _, ok := m1[string(value[j])]; !ok {
				m1[string(value[j])] = string(pat)
			}

			if _, ok := m2[string(pat)]; !ok {
				m2[string(pat)] = string(value[j])
			}

			if m1[string(value[j])] != string(pat) {
				check = false
			}

			if m2[string(pat)] != string(value[j]) {
				check = false
			}
		}
		if check {
			result = append(result, value)
		}
		check = true
	}
	return result
}
